sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SchemaAndRowProto.java

```diff
@@ -0,0 +1,166 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.gcp.bigquery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.cloud.bigquery.v3.RowOuterClass;
+import com.google.cloud.bigquery.v3.RowOuterClass.Row;
+import com.google.cloud.bigquery.v3.RowOuterClass.StructType;
+import com.google.common.collect.ImmutableList;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * A wrapper for a {@link Row} and the {@link StructType} representing the schema of the
+ * table it was generated from.
+ * Also provides function to read the data out.
+ */
+public class SchemaAndRowProto {
+  private final StructType tableSchema;
+  private final Row row;
+  TableRow tableRow;
+
+  public SchemaAndRowProto(StructType tableSchema, Row row) {
+    this.tableSchema = checkNotNull(tableSchema, "tableSchema");
+    this.row = checkNotNull(row, "row");
+  }
+
+  public Row getRow() {
+    return row;
+  }
+
+  public StructType getSchema() {
+    return tableSchema;
+  }
+
+  /**
+   * Gets a field value based on field name.
+   * Note: The implementation of this function will convert the entire row to {@link TableRow}
+   * and use the field name to get the value out of TableRow, based on the assumption that the
+   * most efficient way of accessing the subset of fields is to pass down the Field Selections
+   * to the read API.
+   *
+   * @param fieldName
+   * @return
+   */
+  public Object get(String fieldName) {
+    TableRow currentRow = getTableRow();
+    String[] subFieldNames = fieldName.split(".");
+    for (int i = 0; i < subFieldNames.length; i++) {
+      if (i == subFieldNames.length - 1) {
+        return currentRow.get(subFieldNames[i]);
+      } else {
+        // If there are further nested levels, the current field should contain a TableRow.
+        currentRow = (TableRow) (currentRow.get(subFieldNames[i]));
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Converts a {@class Row} that returned from ParallelRead API to {@class TableRow}
+   * which is a map of {fieldname, value} to be consumed by Data Flow API.
+   */
+  public TableRow getTableRow() {
+    if (tableRow != null) {
```

Comment:
* Why does it need to be cached?
* Do we even need to provide an API returning `TableRow` from v3 reads? It is definitely easier for users to migrate, but I'm not sure we want to encourage perpetuating the horrible performance of `TableRow` in the new API.

Reply:
Hmm -- I can see your point about not wanting to make TableRow easy for new callers to use, but especially for EAP customers, ease of adoption is important. What if we simply removed support for the parallel read path at the level of readTableRows()? We can always remove it down the line in Alpha or Beta.